dlt/_workspace/auth.py

```diff
@@ -0,0 +1,181 @@
+from typing import Optional
+from git import Union
+from pydantic import BaseModel, ValidationError
+import jwt
+
+from dlt._workspace._workspace_context import WorkspaceRunContext
+from dlt._workspace.exceptions import (
+    RuntimeNotAuthenticated,
+    RuntimeOperationNotAuthorised,
+    WorkspaceIdMismatch,
+    LocalWorkspaceIdNotSet,
+    WorkspaceRunContextNotAvailable,
+)
+from dlt._workspace.providers import ProfileConfigTomlProvider
+from dlt._workspace.runtime_clients import API_BASE_URL
+from dlt._workspace.runtime_clients.api.models.me_response import MeResponse
+from dlt.common.configuration.providers.toml import (
+    SECRETS_TOML,
+    ConfigTomlProvider,
+    SecretsTomlProvider,
+)
+
+
+import os
+from tomlkit.toml_file import TOMLFile
+
+from dlt._workspace._workspace_context import WorkspaceRunContext
+from dlt._workspace.exceptions import RuntimeNotAuthenticated
+from dlt._workspace.runtime_clients.api.api.me import me
+from dlt._workspace.runtime_clients.api.client import Client as ApiClient
+
+from dlt.cli.config_toml_writer import WritableConfigValue, write_values
+from dlt.common.configuration.providers.toml import SecretsTomlProvider
+from dlt.common.configuration.specs.pluggable_run_context import RunContextBase
+from dlt.common.configuration.specs.runtime_configuration import RuntimeConfiguration
+
+
+class AuthInfo(BaseModel):
+    user_id: str
+    email: str
+    jwt_token: str
+
+
+class AuthService:
```

Comment:
FYI: Generally speaking we use a more functional programming style for all the cli helpers, you can see this very well in the deploy commands. I'm not sure if this ultimately is good or bad. I'd say you can choose to follow that pattern or choose to create these classes for all the runtime interaction stuff. Just keep in mind that these classes are very short lived and do not retain any state between cli commands, so they are not really that useful. If you want to introduce some kind of dependency injection between these services for better testability, that would be a reason to build it this way. Testing pure functions that only depend on the inputs is a quite nice pattern too though.

Reply:
Yeah, I've noticed - I've implemented the entrypoints in functional style. Chose a class for this one to specifically encapsulate all the auth internals and auth state. It will be very useful in next commands (`dlt runtime deploy` and `dlt runtime run`) - there will be many calls to auth-protected API, so I think keeping this state even for one command is worth it. I could switch to functional style too, it would be a bit more verbose as to my taste, but I've also noticed some internals such as `Provider` and `RunContext` are implemented as classes, so would rather stick to class here if you don't mind?